rfcs/reorganize_src_lib.md

```diff
@@ -0,0 +1,117 @@
+
+## First attempt at grouping
+
+
+I tried to create meaningful groups and to not have to many edges in the following graph of [dependencies between groups](./res/reorganize_src_lib/class_graph.png) (which is still a bit complex).
+
+- The section below contains some information about the groups.
+- In the following dependency graphs, the libraries are colored according to their respective groups. The first one is the most readable (but contains less information) because it does not contain the `Utils` and `Logging` groups, and was run through the `tred` tool, which removes a lot of edges while preserving reachability.
+
+    - [without utils and logging - treded](./res/reorganize_src_lib/no_utils_tred.png)
+    - [without utils and logging](./res/reorganize_src_lib/no_utils.png)
+    - [all libraries - treded](./res/reorganize_src_lib/all_libs_tred.png)
+    - [all libraries](./res/reorganize_src_lib/all_libs.png)
+
+- The exact groups can also be found in the [library_group.md](./res/reorganize_src_lib/library_groups.md) file.
+
+
+
+### Current Groups
+
+- `Others`: Libraries that could stay at the toplevel, as their own group, or that I am not sure about.
+For instance the `mina_base` library that contains a lot a files and will maybe be split in the future.
+
+- `Utils`: Non application specific and usually small libraries. Most of the time leaves of the dependency graph.
+
+- `Logging`: Libraries related to logging that everyone can depend on (even the `Utils` one).
+
+- `Testing`: This folder would would contain two types of testing libraries:
+    - High level calls to tests such as `command_line_tests`.
+    - Test related utils such as `quickcheck_lib` (The difference with the `Utils` folder is that these would be allowed to depend on other application specific libraries).
+
+  The other folders could also contain a `testing` subfolder for specific test related libraries (for instance the `mina_net2_tests` would be in the `Network` group).
+    
+
+- `Processes`: Libraries related to parallelism and inter process communication,
+   such as `child_processes` and `pipe_lib`.
+
+- `Network`: Libraries related to network communication between the various component.
+
+- `Config`: Libraries such as `mina_numbers` defining primitive types used by most other groups.
+
+- `Crypto`:
+  Creating the `crypto` folder was already attempted in [this PR](https://github.com/MinaProtocol/mina/pull/9540) that I used as a reference.
+  Would it make sense to split it between the libraries that depend on the `Config` group,
+  and those that do not ?
+
+- `Snarky`: The libraries from the snarky git submodule.
+
+- `Transition`: Modules related to the handling of transitions
+(many of which have the `transition_` prefix).
+
+- The other name should be self descriptive. I am not very confident about those but I tried to respect clusters in the dependency graph :
+  `Consensus`, `Staged_ledger`, `Genesis_ledger`, `Rosetta` and `Caching`.
```

Comment:
From my side, I would appreciate `Consensus`'s responsibilities being divided a little. There is code for assisting with block production, code for evaluating/validating accepted blocks, code for managing ledgers, and code for managing the 'stake delegation' primitive. These things probably don't need to co-exist in a ~4500 line library.

Reply:
The `Caching` libraries need some reworking, and it would probably make sense to combine them with the logic for loading configuration files (for `Config` libraries) and for storing logs (`Logging`). It would be nice if this wasn't a compile-time dependency, and `Mina_lib` could tie the knot on this instead.